nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java

```diff
@@ -452,243 +305,201 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
-                recordSetWriter.write(record);
+    private boolean determineListable(final FileStatus status, final long minimumAge, final long maximumAge, final boolean isTransitioningFromLegacyState, final boolean isLegacyLastStatusListed) {
+        // If the file was created during the processor's last iteration we have to check if it was already listed
+        // If legacy state was used and the file was already listed once, we don't want to list it once again.
+        if (status.getModificationTime() == fileStatusManager.getLastModificationTime()) {
+            if (isTransitioningFromLegacyState) {
+                return !isLegacyLastStatusListed;
             }
-
-            writeResult = recordSetWriter.finishRecordSet();
+            return !fileStatusManager.getLastModifiedStatuses().contains(status);
         }
 
-        final Map<String, String> attributes = new HashMap<>(writeResult.getAttributes());
-        attributes.put("record.count", String.valueOf(writeResult.getRecordCount()));
-        flowFile = session.putAllAttributes(flowFile, attributes);
+        final long fileAge = System.currentTimeMillis() - status.getModificationTime();
+        if (minimumAge > fileAge || fileAge > maximumAge) {
+            return false;
+        }
 
-        session.transfer(flowFile, getSuccessRelationship());
+        return status.getModificationTime() > fileStatusManager.getLastModificationTime();
     }
 
-    private Record createRecord(final FileStatus fileStatus) {
-        final Map<String, Object> values = new HashMap<>();
-        values.put(FILENAME, fileStatus.getPath().getName());
-        values.put(PATH, getAbsolutePath(fileStatus.getPath().getParent()));
-        values.put(OWNER, fileStatus.getOwner());
-        values.put(GROUP, fileStatus.getGroup());
-        values.put(LAST_MODIFIED, new Timestamp(fileStatus.getModificationTime()));
-        values.put(SIZE, fileStatus.getLen());
-        values.put(REPLICATION, fileStatus.getReplication());
-
-        final FsPermission permission = fileStatus.getPermission();
-        final String perms = getPerms(permission.getUserAction()) + getPerms(permission.getGroupAction()) + getPerms(permission.getOtherAction());
-        values.put(PERMISSIONS, perms);
-
-        values.put(IS_DIRECTORY, fileStatus.isDirectory());
-        values.put(IS_SYM_LINK, fileStatus.isSymlink());
-        values.put(IS_ENCRYPTED, fileStatus.isEncrypted());
-        values.put(IS_ERASURE_CODED, fileStatus.isErasureCoded());
-
-        return new MapRecord(getRecordSchema(), values);
-    }
+    private FileCountRemoteIterator<FileStatus> getFileStatusIterator(final Path path, final boolean recursive, final FileSystem hdfs, final PathFilter filter) {
+        final Deque<Path> pathStack = new ArrayDeque<>();
+        pathStack.push(path);
 
-    private RecordSchema getRecordSchema() {
-        return RECORD_SCHEMA;
-    }
+        return new FileCountRemoteIterator<>() {
```

Comment:
This iterator could be simplified if the stack was handled within and the filtering is not done here.
(`ListHDFS` already does some of in the `determineListable` so all the filtering could be combined in one place.)
Also probably better for it to implement `Iterable` instead of `RemoteIterator`.